aws/src/main/java/org/apache/iceberg/aws/dynamodb/DynamoDbCatalog.java

```diff
@@ -0,0 +1,613 @@
+          .tableName(awsProperties.dynamoDbTableName())
+          .indexName(GSI_NAMESPACE_IDENTIFIER)
+          .keyConditionExpression(condition)
+          .expressionAttributeValues(conditionValues)
+          .build());
+
+      if (response.hasItems()) {
+        for (Map<String, AttributeValue> item : response.items()) {
+          String identifier = item.get(COL_IDENTIFIER).s();
+          if (!COL_IDENTIFIER_NAMESPACE.equals(identifier)) {
+            identifiers.add(TableIdentifier.of(identifier.split("\\.")));
+          }
+        }
+      }
+
+      lastEvaluatedKey = response.lastEvaluatedKey();
+    } while (!lastEvaluatedKey.isEmpty());
+    return identifiers;
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    Map<String, AttributeValue> key = tablePrimaryKey(identifier);
+    try {
+      GetItemResponse response = dynamo.getItem(GetItemRequest.builder()
+          .tableName(awsProperties.dynamoDbTableName())
+          .consistentRead(true)
+          .key(key)
+          .build());
+
+      if (!response.hasItem()) {
+        throw new NoSuchTableException("Cannot find table %s to drop", identifier);
+      }
+
+      TableOperations ops = newTableOps(identifier);
+      TableMetadata lastMetadata = ops.current();
+      dynamo.deleteItem(DeleteItemRequest.builder()
+          .tableName(awsProperties.dynamoDbTableName())
+          .key(tablePrimaryKey(identifier))
+          .conditionExpression(COL_VERSION + " = :v")
+          .expressionAttributeValues(ImmutableMap.of(":v", response.item().get(COL_VERSION)))
+          .build());
+      LOG.info("Successfully dropped table {} from DynamoDb catalog", identifier);
+
+      if (purge && lastMetadata != null) {
+        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+        LOG.info("Table {} data purged", identifier);
+      }
+
+      LOG.info("Dropped table: {}", identifier);
+      return true;
+    } catch (ConditionalCheckFailedException e) {
+      LOG.error("Cannot complete drop table operation for {}: commit conflict", identifier, e);
+      return false;
+    } catch (Exception e) {
+      LOG.error("Cannot complete drop table operation for {}: unexpected exception", identifier, e);
+      throw e;
+    }
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    Map<String, AttributeValue> fromKey = tablePrimaryKey(from);
+    Map<String, AttributeValue> toKey = tablePrimaryKey(to);
+
+    GetItemResponse fromResponse = dynamo.getItem(GetItemRequest.builder()
+        .tableName(awsProperties.dynamoDbTableName())
+        .consistentRead(true)
+        .key(fromKey)
+        .build());
+
+    if (!fromResponse.hasItem()) {
+      throw new NoSuchTableException("Cannot rename table %s to %s: %s does not exist", from, to, from);
+    }
+
+    GetItemResponse toResponse = dynamo.getItem(GetItemRequest.builder()
```

Comment:
This seems like an unnecessary check if it's also being handled by the transaction.

Reply:
Similar to your comment of the version field, a get is necessary to get the last version and perform optimistic locking.

Reply:
Wait, hold on I updated the wrong comment here.  We need to do a get for the row identifying the table to be renamed, but I don't see why we need a get for the target row?  There two gets here.  The first is necessary, the second one shouldn't be from what I can tell.

Reply:
The `toResponse` is only used for the first existence check, but that's also done by the put conditional expression in the transaction.  We shouldn't need that.

Reply:
Yes it is only used to make sure the destination does not exist, because in the transaction block we cannot tell which step it fails at in the 2 step transaction. The early check can provide a clearer message of `AlreadyExistsException` instead of a generic conditional check failure, and it is much more likely that someone accidentally tries to rename to an already existing table, comparing to the case that 1 person tries to create the table and another person tries to rename a table to it at the same time.

Reply:
Hmm, it's unfortunate that the transactions don't have a better mechanism for handling constraint violations (or custom errors messages) for failures.  Since this isn't a high use/frequency operation, I agree the extra check is helpful for users.